egs/wsj/s5/steps/cleanup/combine_short_segments.sh

```diff
@@ -0,0 +1,191 @@
+    chomp;
+    @col = split;
+    $text = join(" ", @col[1..@col-1]);
+    $utt2text{$col[0]} = $text;
+  }
+
+  # Once the under-length segment is found, we concatenate it with
+  # the contiguous segments. We first combine it with segments after it
+  # and then with segments before it.
+  foreach $seg (sort keys %uttlist) {
+    if ($utt2len{$seg} < $min_seg_len && $utt2item{$seg} > 0) {
+      $forward_combining = 0;
+      $backward_combining = 0;
+      @utts = split(" ", $spk2utt{$utt2spk{$seg}});
+      foreach $seg2 (sort @utts) {
+        if ($seg2 gt $seg) {
+          $sum = $utt2len{$seg} + $utt2len{$seg2};
+          $utt2len{$seg} = $sum;
+          $utt2len{$seg2} = -1;
+          $uttlist{$seg} = $uttlist{$seg} . "-" . $uttlist{$seg2};
+          $utt2feat{$seg} = $utt2feat{$seg} . " " . $utt2feat{$seg2};
+          $utt2text{$seg} = $utt2text{$seg} . " " . $utt2text{$seg2};
+          $utt2item{$seg} = $utt2item{$seg} + $utt2item{$seg2};
+          $utt2item{$seg2} = 0;
+          if ($sum >= $min_seg_len) {
+            $forward_combining = 1;
+            last;
+          }
+        }
+      }
+      if ($forward_combining == 0) {
+        foreach $seg2 (reverse sort @utts) {
+          if ($seg2 lt $seg) {
+            $sum = $utt2len{$seg} + $utt2len{$seg2};
+            $utt2len{$seg} = $sum;
+            $utt2len{$seg2} = -1;
+            $uttlist{$seg} = $uttlist{$seg} . "-" . $uttlist{$seg2};
+            $utt2feat{$seg} = $utt2feat{$seg} . " " . $utt2feat{$seg2};
+            $utt2text{$seg} = $utt2text{$seg} . " " . $utt2text{$seg2};
+            $utt2item{$seg} = $utt2item{$seg} + $utt2item{$seg2};
+            $utt2item{$seg2} = 0;
+            if ($sum >= $min_seg_len) {
+              $backward_combining = 1;
+              last;
+            }
+          }
+        }
+      }
+
+      if ($forward_combining == 0 && $backward_combining == 0) { 
+        die "Error: fail to have combined segment with enough length\n"; 
+      }
+    }
+  }
+
+  foreach $seg (sort keys %uttlist) {
+    if ($utt2item{$seg} > 1) {
+      print UO "$uttlist{$seg}-appended $utt2spk{$seg}\n";
+      print FO "$uttlist{$seg}-appended concat-feats $utt2feat{$seg} - |\n";
+      print TO "$uttlist{$seg}-appended $utt2text{$seg}\n";
+    } elsif ($utt2item{$seg} == 1) {
+      print UO "$uttlist{$seg} $utt2spk{$seg}\n";
+      print FO "$uttlist{$seg} $utt2feat{$seg}\n" ;
+      print TO "$uttlist{$seg} $utt2text{$seg}\n";
+    }
+  }
+
+' $input_dir/utt2spk \
+$input_dir/spk2utt \
+$input_dir/text \
+$input_dir/feats.scp \
+$output_dir/feats.length \
+$output_dir/utt2spk \
+$output_dir/text \
+$output_dir/feats.scp
+}
+
+mkdir -p $output_dir
+
+if [ -f $input_dir/segments ]; then
+  awk -v l=$frame_per_sec '{
+    len=($4 - $3) * l;
+    if (len > 0) {
+      printf("%s %d \n", $1, len);
+    }
+  }' $input_dir/segments >$output_dir/feats.length
+else
+  feat-to-len --print-args=false scp:$input_dir/feats.scp ark,t:$output_dir/feats.length
```

Comment:
I suppose to work out the script in the length of frames, so i convert the length in seconds to the length in frames at the beginning of the script. The  frame_shift option is used equivalently.